infer/src/bufferoverrun/bufferOverrunSemantics.ml

```diff
@@ -416,12 +416,27 @@ let rec eval_sympath_partial ~mode params p mem =
       Mem.find_set locs mem
 
 
+(* Import a path p from a callee as a set of locations to a caller.
+   There are the following cases:
+   (1) p involves dereference of a formal parameter:
+     Substitute the formal parameter with actual expression evaluating dereferences:
+     Example: p = "**A.f1", A is a formal parameter, actual is &B, B points
+       to C and D.
+       Return: {"C.f1", "D.f1"}
+   (2) p involves a formal parameter passed by value:
+     Return unknown location.
+     Example: p = A, A is a formal parameter, actual is B
+     Return: {"unknown"}
+     The reason why it is not possible to import A as B is that there is a mapping
+     from formal locations to evaluations of actual expressions in ParamBindings and
+     not to actual expression. That is, if the actual expression is A, we can get its
+     value, not the location itself. This is also why we can import dereferences precisely
+     (we get powloc part of such value). *)
 and eval_locpath ~mode params p mem =
   let res =
     match p with
     | BoField.Prim (Symb.SymbolPath.Pvar _ | Symb.SymbolPath.Callsite _) ->
-        let v = eval_sympath_partial ~mode params p mem in
-        Val.get_all_locs v
+        PowLoc.unknown
```

Comment:
Thank you for the question, the explanation I gave in the commit message was not clear and confusing. The fundamental problem fixed by this commit is indeed just the code in `eval_locpath` referenced above. Importing of alias of return value is only indirectly related to it (see below).

When `p` is `pvar` or `callsite` and we import its value by calling `eval_sympath_partial` then `Val.get_all_locs` returns locations representing possible dereferences of imported `p`, not locations representing imported `p`. On the other hand, when `p` is `deref`, importing the value of dereferenced pointer and calling `Val.get_all_locs` on the result gives locations representing imported `p` (this is the case below).

And indeed, there is no way how to get a location of a variable representing an actual parameter given a location representing a formal parameter as opposed to the case when we have a dereference of  formal parameter. This is because `BufferOverrunSemantics.ParamBindings` is a mapping from formals to evaluation of actuals so we can get a value that was passed to a call, but not the actual expression (from which we could get the variable if the expression is just a variable).

Now, the problem manifested itself when importing an alias bound to a return value, because normally a location representing a formal parameter itself is not imported (as opposed to its value or its dereference).

Here are examples when we cannot precisely import aliases of a return value and BO was importing them incorrectly before the fix:

```c
// alias of return value is x
int* callee_incorrectly_imported_alias (int* x) {
  return x;
}

int test_03 (int* y) {
  if (*(callee_incorrectly_imported_alias (y)) > 0) { // the alias of return value x from callee_incorrectly_imported_alias will be imported as *y
  ...
}

int test_04() {
  int i = 10;
  int* y = &i;
  if (*(callee_incorrectly_imported_alias (y)) > 0) { // the alias of return value x from callee_incorrectly_imported_alias will be imported as i
  ...
}
```

Here is the example when we cannot import an alias of return value precisely and BO was already correctly returning `Loc.unknown` before the fix:

```c
//  alias of return value is x
int callee_no_imported_alias (int x) {
  return x;
}

int test_01 (int y) {
  if (callee_no_imported_alias (y) > 0) {  // x was correctly imported as Loc.unknown
  ...
```

Here is the example, when we can precisely import an alias of return value:

```c
// alias of return value is *x
int callee_imported_alias (int* x) {
  return *x;
}

int test_02 (int y) {
  if (callee_imported_alias (&y) > 0) { // *x is correctly imported as y
```

Reply:
Thank you so much for the explanation and great catch!! :clap::clap::clap::clap::clap: